### ISSUE TITLE
stage3: tuple does not work with funciton pointer

### ISSUE THREAD

**Issue by 20kano** [batiati/mustache-zig#16]
mustache-zig: stage3
zig: 0.11.0-dev.1636+47e14b7ff

```sh
> zig build run
error: _bug...
error: The following command terminated unexpectedly:
```

```zig
const std = @import("std");
const mustache = @import("mustache-zig/src/mustache.zig");

pub const Server = struct {
    handler_fns: std.StringHashMap(HandlerFn),

    pub const HandlerFn = *const fn(Request) void;

    fn handle(self: *Server) !void {
        var req = Request{ .path = "/test"};
        if (self.handler_fns.get(req.path)) |f| f(req);
    }
};

pub const Request = struct {
    path: []const u8 = undefined,
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer std.debug.assert(!gpa.deinit());
    const allocator = gpa.allocator();
    
    var fns = std.StringHashMap(Server.HandlerFn).init(allocator);
    try fns.put("/test", rqTest);
    defer fns.deinit();

    var srv = Server{.handler_fns = fns};
	
    try srv.handle();
}

fn rqTest(req: Request) void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer std.debug.assert(!gpa.deinit());
    const allocator = gpa.allocator();
	
    // const Ctx = struct {name: []const u8};
    // const ctx: Ctx = .{.name = req.path};
    const ctx = .{.name = req.path};
    
    const html = mustache.allocRenderText(allocator, "Hello {{name}} from Zig", ctx) catch return;
    defer allocator.free(html);
    
    std.debug.print("{s}\n", .{html});
}
```
Not sure since I'm new to zig, is this a problem of the zig compiler?

**Comment by batiati** [comment]
Hi there! Nice use case.

Can you elaborate where the function pointer is used/interfere with the mustache template? I couldn't get that part by only reading your code.

Anonymous structs and tuples that contains comptime values (e.g. string literals) are known to trigger a compiler TODO in stage3. Anyway, it doesn't seem to be your case.

**Comment by 20kano** [comment]
Sorry, I'm not sure if the function pointer is the cause, but when I ran it on zig 0.10.1 and the master branch, it worked fine.

In that case, it seems that `|f| f(req);` has been changed in 0.11.0, and as such I wondered if the function pointer was the cause.

**Comment by batiati** [comment]
Humm, I see.
The stage2 compiler notation for function pointers is `*const fn(....)` and the `fn(...)` form is for comptime functions.
On stage1 (zig 0.10.x and earlier) both concepts were represented by `fn(...)`.

For mustache-zig, if you are using Zig 0.11.0-dev, you should use a branch named `stage3` which isn't ready yet but can cover most of the uses cases (except some comptime templates and a few other edge cases).

(...)

Nothing related to zig or mustache, but by reading your code again, I spotted something that can be an issue:

```zig
pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer std.debug.assert(!gpa.deinit());
    const allocator = gpa.allocator();
    
   // You're creating the hashmap here
    var fns = std.StringHashMap(Server.HandlerFn).init(allocator);
    try fns.put("/test", rqTest);
    defer fns.deinit();

    // And them copying it to the server
    var srv = Server{.handler_fns = fns};
	
    try srv.handle();
}
```

I'd suggest you either make `Server.handler_fns` a reference or implement `Server.init()` to initialize it,

Like this:

```zig
   // Option #1
    var fns = std.StringHashMap(Server.HandlerFn).init(allocator);
    try fns.put("/test", rqTest);
    defer fns.deinit();

    var srv = Server{.handler_fns = &fns};

```

Or this

```zig
    // Option #2
    var srv = try Server.init(allocator);
    defer srv.deinit();

    try server.handler_fns.put("/test", rqTest);
```

I'm closing this one, let's reserve the GH issue tracker only for mustache-related issues, nevertheless, feel free to add me on Discord (batiati#5408) for any doubts.

**Comment by 20kano** [comment]
I tried the fixes you suggested, but the results were the same.

Humm, this works
```
const Ctx = struct {name: []const u8}
const ctx: Ctx = . {.name = req.path};
```
But strange, this does not, even give an error message on compilation.

```
const ctx = .{.name = req.path};
```

**Comment by 20kano** [comment]
https://github.com/batiati/mustache-zig/blob/6f3cc00cca1465fe761881a0397d408b610caa17/src/rendering/rendering.zig#L970

I could tell that the program was stumbling here, but I could not investigate beyond that point with my level of zig experience.

**Comment by batiati** [comment]
> could tell that the program was stumbling here, but I could not investigate beyond that point with my level of zig experience.

I've experienced the same issue too, in many places I had to change from anonymous structs and tuples to fully declared structs and arrays. I guess it is just a compiler issue, but I'm not paying much attention to that because stage2 is still under heavy development.

I'm willing to hold this project until Zig releases `0.11.0`, and then we can dig deeper.
